File: client/src-tauri/src/model/defect.rs
use crate::db::table::HasKey;
use crate::model::default_value::DefaultValue;
use crate::model::macro_model::model_with_default;
use crate::model::tag::Tag;
use crate::model::time_stamps::Timestamps;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub enum DefectSeverity {
    Minor,    // 軽微
    Major,    // 重要
    Critical, // 致命的
    Blocker,  // 作業不能
}

impl DefaultValue for DefectSeverity {
    fn default_value() -> Self {
        DefectSeverity::Minor
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
#[serde(rename_all = "PascalCase")]
pub enum DefectStatus {
    Open,       // 登録
    InProgress, // 修正中
    Fixed,      // 修正完了
    Verified,   // 確認完了
    Closed,     // 完全クローズ
}

impl DefaultValue for DefectStatus {
    fn default_value() -> Self {
        DefectStatus::Open
    }
}

impl HasKey<(String, String)> for Defect {
    fn key(&self) -> (String, String) {
        (self.project_id.clone(), self.id.clone())
    }
}

model_with_default!(
    Defect,
    DefectRequest,
    {
        id: String,
        project_id: String,
    },
    {
        task_id: Option<String>,
        title: String,
        description: String,
        severity: DefectSeverity,
        status: DefectStatus,
        owner: String,
        reviewer: String,
        due_date: Option<String>,
        fixed_date: Option<String>,
        verified_date: Option<String>,
        tags: Vec<Tag>,
    }
);

File: client/src/types/db/defect.ts
import { Tag, Timestamps } from "./common";

export enum DefectStatus {
    Open = "Open",       // 登録
    InProgress = "InProgress", // 修正中
    Fixed = "Fixed",      // 修正完了
    Verified = "Verified",   // 確認完了
    Closed = "Closed",     // 完全クローズ
}

export enum DefectSeverity {
    Minor = "Minor",
    Major = "Major",
    Critical = "Critical",
    Blocker = "Blocker",
}


export type DefectKey = {
    id: string,
    project_id: string,
};

export type DefectValue = {
    task_id: string,
    title: string,
    description: string,
    severity: DefectSeverity,
    status: DefectStatus,
    owner: string,
    reviewer: string,
    due_date: string,
    fixed_date: string,
    verified_date: string,
    tags: Tag[],
    timestamps: Timestamps,
};

// Row = Key + Value
export type DefectRow = DefectKey & DefectValue;
export type DefectPayload = DefectKey & Partial<DefectValue>;

export const toDefectPayload = (param: DefectRow): DefectPayload => (
    {
        id: param.id,
        project_id: param.project_id,
        task_id: param.task_id,
        title: param.title,
        description: param.description,
        severity: param.severity,
        status: param.status,
        owner: param.owner,
        reviewer: param.reviewer,
        due_date: param.due_date,
        fixed_date: param.fixed_date,
        verified_date: param.verified_date,
        tags: param.tags,
    }
)
